pkg/validation/spec/gnostic.go

```diff
@@ -0,0 +1,1554 @@
+	}
+
+	minLen := g.GetMinLength()
+	if minLen != 0 {
+		k.MinLength = &minLen
+	}
+
+	k.Pattern = g.GetPattern()
+
+	maxItems := g.GetMaxItems()
+	if maxItems != 0 {
+		k.MaxItems = &maxItems
+	}
+
+	minItems := g.GetMinItems()
+	if minItems != 0 {
+		k.MinItems = &minItems
+	}
+
+	k.UniqueItems = g.GetUniqueItems()
+
+	multOf := g.GetMultipleOf()
+	if multOf != 0 {
+		k.MultipleOf = &multOf
+	}
+
+	enums := g.GetEnum()
+
+	if enums != nil {
+		k.Enum = make([]interface{}, len(enums))
+		for i, v := range enums {
+			var convert interface{}
+			if err := v.ToRawInfo().Decode(&convert); err != nil {
+				return false, err
+			} else {
+				k.Enum[i] = convert
+			}
+		}
+	}
+
+	return true, nil
+}
+
+type GnosticSimpleSchema interface {
+	GetType() string
+	GetFormat() string
+	GetItems() *openapi_v2.PrimitivesItems
+	GetCollectionFormat() string
+	GetDefault() *openapi_v2.Any
+}
+
+func (k *SimpleSchema) FromGnostic(g GnosticSimpleSchema) (ok bool, err error) {
+	ok = true
+
+	if g == nil {
+		return true, nil
+	}
+
+	k.Type = g.GetType()
+	k.Format = g.GetFormat()
+	k.CollectionFormat = g.GetCollectionFormat()
+
+	items := g.GetItems()
+	if items != nil {
+		k.Items = &Items{}
+		if nok, err := k.Items.FromGnostic(items); err != nil {
+			return false, err
+		} else if !nok {
+			ok = false
+		}
+	}
+
+	def := g.GetDefault()
+	if def != nil {
+		var convert interface{}
+		if err := def.ToRawInfo().Decode(&convert); err != nil {
+			return false, err
+		} else {
+			k.Default = convert
+		}
+	}
+
+	return ok, nil
+}
+
+func (k *Items) FromGnostic(g *openapi_v2.PrimitivesItems) (ok bool, err error) {
+	ok = true
+
+	if g == nil {
+		return true, nil
+	}
+
+	if nok, err := k.SimpleSchema.FromGnostic(g); err != nil {
+		return false, err
+	} else if !nok {
+		ok = false
+	}
+
+	if nok, err := k.CommonValidations.FromGnostic(g); err != nil {
+		return false, err
+	} else if !nok {
+		ok = false
+	}
+
+	if nok, err := k.VendorExtensible.FromGnostic(g.VendorExtension); err != nil {
+		return false, err
+	} else if !nok {
+		ok = false
+	}
+
+	return ok, nil
+}
+
+func (k *VendorExtensible) FromGnostic(g []*openapi_v2.NamedAny) (ok bool, err error) {
+	ok = true
+	if len(g) == 0 {
+		return true, nil
+	}
+
+	k.Extensions = make(Extensions, len(g))
+	for _, v := range g {
+		if v.Value == nil {
+			k.Extensions[v.Name] = nil
+			continue
+		}
+
+		var iface interface{}
+		if err := v.Value.ToRawInfo().Decode(&iface); err != nil {
+			return false, err
+		} else {
+			k.Extensions[v.Name] = iface
+		}
+	}
+	return ok, nil
+}
+
+func (k *Refable) FromGnostic(g string) (ok bool, err error) {
+	return k.Ref.FromGnostic(g)
+}
+
+func (k *Ref) FromGnostic(g string) (ok bool, err error) {
+	if g == "" {
+		return true, nil
+	}
+
+	ref, err := jsonreference.New(g)
+	if err != nil {
+		return false, err
+	}
+
+	*k = Ref{
+		Ref: ref,
+	}
+
+	return true, nil
+}
+
+// Swagger Document
+
+func (k *Swagger) FromGnostic(g *openapi_v2.Document) (ok bool, err error) {
```

Comment:
document the meaning of the ok and err returns (`!ok` means a value we know we can't round-trip was present in the gnostic document, err means an unexpected error occurred converting)

Reply:
also document that zero-value validation constraints like maxItems/maxLength/maxProperties are lost via gnostic

Reply:
added